### PR TITLE
Enable Boogie programs to be deeply nested without overflowing the stack

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Changes
- Do not create threads with 16MB instead of the default 1 MB stack space.
- Stop using a TaskScheduler that always creates a new Thread to run a task.
- Remove a deadlock cause in `SMTLibProcess` that was caused by `TaskCompletionSource.SetResult` call inside a lock where the continuation could try to lock the same monitor.
- Stop synchronously blocking on Tasks inside an asynchronous context, which can cause deadlocks
- Wasn't needed: replace algorithms in Boogie that traverse Boogie programs using the call stack

### Testing
- Add a test that tries to trigger a stack overflow by using a Boogie program with deep nesting
- Dafny tests pass with this change: https://github.com/dafny-lang/dafny/pull/1917

### CODE REVIEW

**Comment by cpitclaudel** [comment]
Seems reasonable to me, though I'd prefer to get some feedback from other Boogie users.  Why was this introduced in the first place?

**Comment by keyboardDrummer** [comment]
> Seems reasonable to me, though I'd prefer to get some feedback from other Boogie users. Why was this introduced in the first place?

I contacted @BarryBo who contributed the initial change that added this stack, who informed me this was needed by Dafny. I've run the Dafny test suite and one large Dafny customer test suite with this change and both pass so it seems Dafny does not need this change at the moment.

**Comment by cpitclaudel** [comment]
Very nice, thanks.

**Review by cpitclaudel — DISMISSED** [review]

File: Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs [hunk 1]
```diff
@@ -50,9 +53,10 @@ class SubWriter : WriterWrapper {
     }
 
     public string SetTargetAndGetBuffer(TextWriter newTarget) {
-      var result = buffering ? ((StringWriter)target).ToString() : "";
-      buffering = false;
+      var oldTarget = target;
       target = newTarget;
+      var result = buffering ? ((StringWriter)oldTarget).ToString() : "";
+      buffering = false;
       return result;
```
Comment (atomb) [inline Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs:60]: Nice catch. This is a subtle issue!

File: Source/Provers/SMTLib/SMTLibProcess.cs [hunk 2]
```diff
@@ -11,6 +11,8 @@ namespace Microsoft.Boogie.SMTLib
 {
   /*
    * Not thread-safe.
+   * The locks inside this class serve to synchronize between the single external user and the
+   * internal IO threads.
```
Comment (atomb) [inline Source/Provers/SMTLib/SMTLibProcess.cs:15]: This is nice to have documented.

File: Source/VCGeneration/SplitAndVerifyWorker.cs [hunk 3]
```diff
@@ -160,6 +165,65 @@ private async Task ProcessResult(Split split, CancellationToken cancellationToke
...
+          {
+            return Outcome.Inconclusive;
+          }
+          return currentOutcome;
+        default:
+          Contract.Assert(false);
+          throw new cce.UnreachableException();
+      }
+    }
+
```
Comment (atomb) [inline Source/VCGeneration/SplitAndVerifyWorker.cs:226]: Thanks for moving these to separate methods. I think I'll want to use them to resolve #494.

File: Source/Houdini/StagedHoudini.cs [hunk 4]
```diff
@@ -163,8 +163,7 @@ public HoudiniOutcome PerformStagedHoudiniInference()
       foreach (var s in plan)
       {
         Debug.Assert(!plan.GetDependences(s).Contains(s));
-        tasks.Add(new StagedHoudiniTask(s,
-          new Task(o => { ExecuteStage((ScheduledStage) o); }, s, TaskCreationOptions.LongRunning)));
+        tasks.Add(new StagedHoudiniTask(s, Task.Run(() => ExecuteStage(s))));
```
Comment (atomb) [inline Source/Houdini/StagedHoudini.cs:166]: Will this new task still have the `LongRunning` option? Does it matter?

File: Source/ExecutionEngine/ExecutionEngine.cs [hunk 5]
```diff
@@ -236,23 +236,26 @@ public class ExecutionEngine : IDisposable {
...
+        foreach (var fileName in fileNames) {
+          success &= await ProcessFiles(output, new List<string> { fileName }, lookForSnapshots, fileName);
+        }
+        return success;
       }
 
       if (0 <= Options.VerifySnapshots && lookForSnapshots) {
         var snapshotsByVersion = LookForSnapshots(fileNames);
-        return snapshotsByVersion.All(s => {
-          // BUG: Reusing checkers during snapshots doesn't work, even though it should. We create a new engine (and thus checker pool) to workaround this.
```
Comment (atomb) [inline Source/ExecutionEngine/ExecutionEngine.cs:252]: Does this bug still exist?

**Review by atomb — DISMISSED** [review]
Nice improvement! I don't think any of my comments necessarily require any changes, but I'll leave that to your judgement.

File: Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs (hunk 1 above)
Comment (MikaelMayer) [reply Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs:60]: Great catch. I'm wondering if it is still possible to have another race condition however: Two threads stepping in this code could both assign `oldTarget` together, and then assign `target` together, and one StringWriter would not be rendered.
I think the whole function should be locked someway.

Comment (keyboardDrummer) [reply Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs:60]: The calls to this method are inside a lock

**Review by atomb — APPROVED** [review]

File: Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs (hunk 1 above)
Comment (cpitclaudel) [reply Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs:60]: I think the "right" way to solve these sort of issues is `Interlocked.Exchange`, but I don't actually understand the root cause, here.

If this method is called within an exclusive lock, how can this change fix an issue?  And if not, why not just lock the whole method body on `myLock`?

Comment (keyboardDrummer) [reply Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs:60]: The issue was that we had:
```
var result = buffering ? ((StringWriter)target).ToString() : "";
...
target = newTarget;
```

So something could be written to `target` after the `var result` line but before the `target = newTarget` line.

Comment (cpitclaudel) [reply Source/ExecutionEngine/ConcurrentToSequentialWriteManager.cs:60]: Yes but do you have a guarantee that writes to these two variables result and target are done in order?  Otherwise the fix is not guaranteed to help, right?
